### ISSUE TITLE
URIs containing unicode characters cause validation error

### ISSUE THREAD

**Issue by xplosionmind** [aarongustafson/jekyll-webmention_io#160]
My Netlify build log (full log [here](https://app.netlify.com/sites/xplosionmind/deploys/6194382ccd88b035d0d4ca8e)):
```
12:04:10 AM: $ JEKYLL_ENV=production jekyll build
12:04:11 AM: NOTE: nokogumbo: Using Nokogiri::HTML5 provided by Nokogiri. See https://github.com/sparklemotion/nokogiri/issues/2205 for more information.
12:04:11 AM: Configuration file: /opt/build/repo/_config.yml
12:04:11 AM:             Source: /opt/build/repo
12:04:11 AM:        Destination: /opt/build/repo/_site
12:04:11 AM:  Incremental build: disabled. Enable with --incremental
12:04:11 AM:       Generating...
12:04:11 AM: [jekyll-webmention_io] Beginning to gather webmentions of your posts. This may take a while.
12:04:16 AM:                     ------------------------------------------------
12:04:16 AM:       Jekyll 4.2.1   Please append `--trace` to the `build` command 
12:04:16 AM:                      for any additional information or backtrace. 
12:04:16 AM:                     ------------------------------------------------
12:04:17 AM: /opt/buildhome/.rvm/rubies/ruby-3.0.2/lib/ruby/3.0.0/uri/rfc2396_parser.rb:176:in `split': bad URI(is not URI?): https://webmention.io/api/mentions?target[]=https://tommi.space/Maturandi&target[]=https://tommi.space/maturità&target[]=https://tommi.space/maturita&target[]=https://www.tommi.space/Maturandi&target[]=http://tommasomarmo.com/Maturandi&sort-by=published&perPage=9999 (URI::InvalidURIError)
	from /opt/buildhome/.rvm/rubies/ruby-3.0.2/lib/ruby/3.0.0/uri/rfc2396_parser.rb:210:in `parse'
	from /opt/build/cache/bundle/ruby/3.0.0/bundler/gems/jekyll-webmention_io-07525ac025bf/lib/jekyll/webmention_io.rb:366:in `uri_ok?'
	from /opt/build/cache/bundle/ruby/3.0.0/bundler/gems/jekyll-webmention_io-07525ac025bf/lib/jekyll/webmention_io.rb:280:in `get_uri_source'
	from /opt/build/cache/bundle/ruby/3.0.0/bundler/gems/jekyll-webmention_io-07525ac025bf/lib/jekyll/webmention_io.rb:131:in `get_response'
	from /opt/build/cache/bundle/ruby/3.0.0/bundler/gems/jekyll-webmention_io-07525ac025bf/lib/jekyll/generators/gather_webmentions.rb:89:in `check_for_webmentions'
	from /opt/build/cache/bundle/ruby/3.0.0/bundler/gems/jekyll-webmention_io-07525ac025bf/lib/jekyll/generators/gather_webmentions.rb:52:in `block in generate'
	from /opt/build/cache/bundle/ruby/3.0.0/bundler/gems/jekyll-webmention_io-07525ac025bf/lib/jekyll/generators/gather_webmentions.rb:51:in `each'
	from /opt/build/cache/bundle/ruby/3.0.0/bundler/gems/jekyll-webmention_io-07525ac025bf/lib/jekyll/generators/gather_webmentions.rb:51:in `generate'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/site.rb:193:in `block in generate'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/site.rb:191:in `each'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/site.rb:191:in `generate'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/profiler.rb:43:in `block in profile_process'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/profiler.rb:41:in `each'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/profiler.rb:41:in `profile_process'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/site.rb:75:in `process'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/command.rb:28:in `process_site'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/commands/build.rb:65:in `build'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/commands/build.rb:36:in `process'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/command.rb:91:in `block in process_with_graceful_fail'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/command.rb:91:in `each'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/command.rb:91:in `process_with_graceful_fail'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/commands/build.rb:18:in `block (2 levels) in init_with_program'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `block in execute'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `each'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `execute'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/mercenary-0.4.0/lib/mercenary/program.rb:44:in `go'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/mercenary-0.4.0/lib/mercenary.rb:21:in `program'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/exe/jekyll:15:in `<top (required)>'
	from /opt/build/cache/binstubs/jekyll:29:in `load'
	from /opt/build/cache/binstubs/jekyll:29:in `<main>'
​```

**Comment by fancypantalons** [comment]
That's... a very weird looking, clearly broken URI.

This looks like a possible interaction with the redirect-from plugin.  

Could you post the frontmatter for the page where the webmentions are originating?  Judging by the list of URLs, I expect it's a page called "Maturandi".

**Comment by xplosionmind** [comment]
> Could you post the frontmatter for the page where the webmentions are originating?

Sure. Thanks a lot for your help in troubleshooting.  
You can see the source code of the page [here](https://github.com/xplosionmind/tommi.space/blob/main/_posts/2019-06-16-Maturandi.md), while its front matter is here:
```yaml
date: 2019-06-16
image: "/images/posts/2019-06-16-maturandi.jpg"
redirect_from: ["/maturità", "/maturita"]
```

**Comment by fancypantalons** [comment]
Huh, do you know where these URLs are coming from?

- http://tommasomarmo.com/Maturandi
- https://www.tommi.space/Maturandi

I checked the configuration you threw in #147 and I didn't see reference to either of these domains, and yet they're appearing in the list of collected URIs during the webmention pass, so they've gotta be coming from somewhere...

**Comment by xplosionmind** [comment]
> Huh, do you know where these URLs are coming from?

Those are URLs I set in [my `_config.yml`](https://github.com/xplosionmind/tommi.space/blob/b4413746098190940ab304cb7ad6b735d5be063b/_config.yml#L139) as `legacy_domains`:
```
webmentions:
  username: tommi.space
  cache_bad_uris_for: 5
  legacy_domains:
    - https://www.tommi.space
    - http://tommasomarmo.com
```

**Comment by fancypantalons** [comment]
Ah.  Yup, that'd do it.  I suspect the code is failing to escape those URIs.  I'll dig around.  Thanks for the additional information!

**Comment by xplosionmind** [comment]
>  that'd do it.

Unfortunately, no.
As you can see below, there is still a problem of compatibility with [jekyll-redirect-from plugin](https://github.com/jekyll/jekyll-redirect-from):

```
12:29:20 AM: $ JEKYLL_ENV=production jekyll build
12:29:22 AM: NOTE: nokogumbo: Using Nokogiri::HTML5 provided by Nokogiri. See https://github.com/sparklemotion/nokogiri/issues/2205 for more information.
12:29:22 AM: Configuration file: /opt/build/repo/_config.yml
12:29:22 AM:             Source: /opt/build/repo
12:29:22 AM:        Destination: /opt/build/repo/_site
12:29:22 AM:  Incremental build: disabled. Enable with --incremental
12:29:22 AM:       Generating...
12:29:23 AM: [jekyll-webmention_io] Beginning to gather webmentions of your posts. This may take a while.
12:29:28 AM:                     ------------------------------------------------
12:29:28 AM:       Jekyll 4.2.1   Please append `--trace` to the `build` command 
12:29:28 AM:                      for any additional information or backtrace. 
12:29:28 AM:                     ------------------------------------------------
12:29:28 AM: /opt/buildhome/.rvm/rubies/ruby-3.0.2/lib/ruby/3.0.0/uri/rfc2396_parser.rb:176:in `split': bad URI(is not URI?): https://webmention.io/api/mentions?target[]=https://tommi.space/Maturandi&target[]=https://tommi.space/maturità&target[]=https://tommi.space/maturita&sort-by=published&perPage=9999 (URI::InvalidURIError)
	from /opt/buildhome/.rvm/rubies/ruby-3.0.2/lib/ruby/3.0.0/uri/rfc2396_parser.rb:210:in `parse'
	from /opt/build/cache/bundle/ruby/3.0.0/bundler/gems/jekyll-webmention_io-07525ac025bf/lib/jekyll/webmention_io.rb:366:in `uri_ok?'
	from /opt/build/cache/bundle/ruby/3.0.0/bundler/gems/jekyll-webmention_io-07525ac025bf/lib/jekyll/webmention_io.rb:280:in `get_uri_source'
	from /opt/build/cache/bundle/ruby/3.0.0/bundler/gems/jekyll-webmention_io-07525ac025bf/lib/jekyll/webmention_io.rb:131:in `get_response'
	from /opt/build/cache/bundle/ruby/3.0.0/bundler/gems/jekyll-webmention_io-07525ac025bf/lib/jekyll/generators/gather_webmentions.rb:89:in `check_for_webmentions'
	from /opt/build/cache/bundle/ruby/3.0.0/bundler/gems/jekyll-webmention_io-07525ac025bf/lib/jekyll/generators/gather_webmentions.rb:52:in `block in generate'
	from /opt/build/cache/bundle/ruby/3.0.0/bundler/gems/jekyll-webmention_io-07525ac025bf/lib/jekyll/generators/gather_webmentions.rb:51:in `each'
	from /opt/build/cache/bundle/ruby/3.0.0/bundler/gems/jekyll-webmention_io-07525ac025bf/lib/jekyll/generators/gather_webmentions.rb:51:in `generate'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/site.rb:193:in `block in generate'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/site.rb:191:in `each'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/site.rb:191:in `generate'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/profiler.rb:43:in `block in profile_process'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/profiler.rb:41:in `each'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/profiler.rb:41:in `profile_process'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/site.rb:75:in `process'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/command.rb:28:in `process_site'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/commands/build.rb:65:in `build'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/commands/build.rb:36:in `process'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/command.rb:91:in `block in process_with_graceful_fail'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/command.rb:91:in `each'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/command.rb:91:in `process_with_graceful_fail'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/lib/jekyll/commands/build.rb:18:in `block (2 levels) in init_with_program'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `block in execute'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `each'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `execute'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/mercenary-0.4.0/lib/mercenary/program.rb:44:in `go'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/mercenary-0.4.0/lib/mercenary.rb:21:in `program'
	from /opt/build/cache/bundle/ruby/3.0.0/gems/jekyll-4.2.1/exe/jekyll:15:in `<top (required)>'
	from /opt/build/cache/binstubs/jekyll:29:in `load'
	from /opt/build/cache/binstubs/jekyll:29:in `<main>'
```

[full log](https://app.netlify.com/teams/xplosionmind/builds/61958fc33959060007cf6cd1)

**Comment by fancypantalons** [comment]
> Unfortunately, no.
> As you can see below, there is still a problem of compatibility with [jekyll-redirect-from plugin](https://github.com/jekyll/jekyll-redirect-from):

Heh, that's a misunderstanding, I haven't fixed anything yet!  :smile:  I'll likely take a look at this issue this weekend.

**Comment by fancypantalons** [comment]
Alright @xplosionmind , I managed to replicate this by adding a `legacy_domains` block and including one of your domain names that includes a unicode character.

I've pushed a change to the `main` branch that fixes this by ensuring we escape the webmention URI before we fire it.

With my test configuration that resolved the issue.

I'll leave this ticket open until you've had a chance to test the change for yourself.
